rclpy/src/rclpy/serialization.cpp

```diff
@@ -0,0 +1,149 @@
+#include <stdexcept>
+
+extern "C"
+{
+#include "rclpy_common/common.h"
+}
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+
+#include "serialization.hpp"
+
+namespace rclpy
+{
+class SerializedMessage
+{
+public:
+  SerializedMessage()
+  {
+    rcl_msg_ = rmw_get_zero_initialized_serialized_message();
+  }
+
+  void
+  init(rcutils_allocator_t allocator)
+  {
+    rcutils_ret_t rcutils_ret = rmw_serialized_message_init(&rcl_msg_, 0u, &allocator);
+    if (RCUTILS_RET_OK != rcutils_ret) {
+      throw RCLError("failed to initialize serialized message");
+    }
+  }
+
+  ~SerializedMessage()
+  {
+    rcutils_ret_t ret = rmw_serialized_message_fini(&rcl_msg_);
+    if (RCUTILS_RET_OK != ret) {
+      fprintf(
+        stderr,
+        "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+        "failed to fini rcl_serialized_msg_t (%d) in destructor: %s\n",
+        ret,
+        rcutils_get_error_string().str);
+      rcutils_reset_error();
+    }
+  }
+
+  rcl_serialized_message_t rcl_msg_;
+};
+
+py::bytes
+serialize(py::object pymsg, py::object pymsg_type)
+{
+  // Get type support
+  auto ts = static_cast<rosidl_message_type_support_t *>(
+    rclpy_common_get_type_support(pymsg_type.ptr()));
+  if (!ts) {
+    throw py::error_already_set();
+  }
+
+  destroy_ros_message_signature * destroy_ros_message = NULL;
+  auto deleter = [&](void * ptr) {destroy_ros_message(ptr);};
+  auto ros_msg = std::unique_ptr<void, decltype(deleter)>(
+    rclpy_convert_from_py(pymsg.ptr(), &destroy_ros_message), deleter);
+  if (!ros_msg) {
+    throw py::error_already_set();
+  }
+
+  // Create a serialized message object
+  SerializedMessage serialized_msg;
+  serialized_msg.init(rcutils_get_default_allocator());
+
+  // Serialize
+  rmw_ret_t rmw_ret = rmw_serialize(ros_msg.get(), ts, &serialized_msg.rcl_msg_);
+  if (RMW_RET_OK != rmw_ret) {
+    throw RCLError("Failed to serialize ROS message");
+  }
+
+  // Bundle serialized message in a bytes object
+  return py::bytes(
+    reinterpret_cast<const char *>(serialized_msg.rcl_msg_.buffer),
+    serialized_msg.rcl_msg_.buffer_length);
+}
+
+py::object
+deserialize(py::bytes pybuffer, py::object pymsg_type)
+{
+  // Get type support
+  auto ts = static_cast<rosidl_message_type_support_t *>(
+    rclpy_common_get_type_support(pymsg_type.ptr()));
+  if (!ts) {
+    throw py::error_already_set();
+  }
+
+  // Create a serialized message object
+  rcl_serialized_message_t serialized_msg = rmw_get_zero_initialized_serialized_message();
+  // Just copy pointer to avoid extra allocation and copy
+  char * serialized_buffer;
+  ssize_t length;
+  if (PYBIND11_BYTES_AS_STRING_AND_SIZE(pybuffer.ptr(), &serialized_buffer, &length)) {
+    throw std::runtime_error("Failed to get py::bytes data");
+  }
+  serialized_msg.buffer_capacity = length;
+  serialized_msg.buffer_length = length;
+  serialized_msg.buffer = reinterpret_cast<uint8_t *>(serialized_buffer);
```

Comment:
@sloretz nit: why not using `serialized_msg` members in place instead of having these `serialized_buffer` and `length` temporary variables?

Reply:
Unsigned/signed conversion - speaking of which I added a check that length is greater than 0 before casting in 9efb9cc.